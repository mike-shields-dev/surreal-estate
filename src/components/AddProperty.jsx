import React, { useState } from "react";
import styles from "../styles/AddProperty.module.css";
import ComboBox from "./ComboBox";
import SpinButton from "./SpinButton";
import TextBox from "./TextBox";
import Alert from "./Alert";
import addProperty from "../requests/addProperty";

const cities = ["Manchester", "Leeds", "Sheffield", "Liverpool"];
const types = [
  "Flat",
  "Semi-detached",
  "Terraced",
  "End of Terrace",
  "Cottage",
  "Bungalow",
];

const initialState = {
  alert: {
    message: "",
    isSuccess: false,
  },
  fields: {
    title: "",
    city: cities[0],
    type: types[0],
    bedrooms: 1,
    bathrooms: 1,
    price: 0,
    email: "",
  },
};

const AddProperty = () => {
  const [fields, setFields] = useState(initialState.fields);
  const [alert, setAlert] = useState(initialState.alert);

  const handleAddProperty = async (event) => {
    event.preventDefault();
    const isPropertyAdded = await addProperty(fields);
    setAlert({
      message: isPropertyAdded
        ? "Property added"
        : "Something went wrong, please try again later",
      isSuccess: isPropertyAdded,
    });
  };

  const handleFieldChange = (event) => {
    const { name, value } = event.target;
    setFields({ ...fields, [name]: value });
  };

  return (
    <>
      <p>Add Property Page</p>
      <div className={styles["add-property"]}>
        <Alert {...alert} />
        <form
          className={styles["add-property__form"]}
          name="add property"
          onSubmit={handleAddProperty}
        >
          <TextBox
            label="Title"
            name="title"
            onChange={handleFieldChange}
            type="text"
            value={fields.title}
          />
          <ComboBox
            label="Type"
            name="type"
            onChange={handleFieldChange}
            value={fields.type}
            values={types}
          />
          <SpinButton
            label="Bedrooms"
            min={0}
            name="bedrooms"
            onChange={handleFieldChange}
            step={1}
            value={fields.bedrooms}
          />
          <SpinButton
            label="Bathrooms"
            min={0}
            name="bathrooms"
            onChange={handleFieldChange}
            step={1}
            value={fields.bathrooms}
          />
          <SpinButton
            label="Price Â£"
            min={0}
            name="price"
            onChange={handleFieldChange}
            step={5000}
            value={fields.price}
          />
          <ComboBox
            label="City"
            name="city"
            onChange={handleFieldChange}
            value={fields.city}
            values={cities}
          />
          <TextBox
            label="Email"
            name="email"
            onChange={handleFieldChange}
            value={fields.email}
            type="email"
          />
          <button type="submit">Add</button>
        </form>
      </div>
    </>
  );
};

export default AddProperty;
