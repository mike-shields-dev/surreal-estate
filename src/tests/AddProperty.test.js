import React from "react";
import { render, screen } from "@testing-library/react";
import AddProperty from "../components/AddProperty";

const renderAddProperty = () => render(<AddProperty />);

describe("AddProperty", () => {
  it("matches snapshot", () => {
    const { asFragment } = renderAddProperty();

    expect(asFragment()).toMatchSnapshot();
  });

  it("renders a form element", () => {
    renderAddProperty();

    expect(screen.getByRole("form")).toBeInTheDocument();
  });

  it("renders a textbox for the 'title' field", () => {
    renderAddProperty();

    expect(screen.getByRole("textbox", { name: /title/i })).toBeInTheDocument();
  });

  it("renders a combobox for the 'type' field", () => {
    renderAddProperty();

    expect(screen.getByRole("combobox", { name: /type/i })).toBeInTheDocument();
  });

  it("renders a spinbutton for the 'bedrooms' field", () => {
    renderAddProperty();

    expect(
      screen.getByRole("spinbutton", { name: /bedrooms/i })
    ).toBeInTheDocument();
  });

  it("renders a spinbutton for the 'bathrooms' field", () => {
    renderAddProperty();

    expect(
      screen.getByRole("spinbutton", { name: /bathrooms/i })
    ).toBeInTheDocument();
  });

  it("renders a spinbutton for the 'price' field", () => {
    renderAddProperty();

    expect(
      screen.getByRole("spinbutton", { name: /price/i })
    ).toBeInTheDocument();
  });

  it("renders a combobox for the 'city' field", () => {
    renderAddProperty();

    expect(screen.getByRole("combobox", { name: /city/i })).toBeInTheDocument();
  });

  it("renders a email input for the 'email' field", () => {
    renderAddProperty();

    expect(screen.getByRole("textbox", { name: /email/i })).toBeInTheDocument();
  });

  it("renders a form submit button", () => {
    renderAddProperty();

    expect(
      screen.getByRole("button", { target: { type: "submit", name: /add/i } })
    ).toBeInTheDocument();
  });
});
