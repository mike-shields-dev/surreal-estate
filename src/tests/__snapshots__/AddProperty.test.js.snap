// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`AddProperty matches snapshot 1`] = `
<DocumentFragment>
  <div
    class="add-property"
    title="add property"
  >
    <form
      class="add-property__form"
      name="add property"
    >
      <div
        class="textbox"
      >
        <label
          class="textbox__label"
          for="title"
        >
          title
        </label>
        <input
          id="title"
          maxlength="48"
          name="title"
          required=""
          type="text"
          value=""
        />
      </div>
      <label
        for="type"
      >
        type
        <select
          id="type"
          name="type"
          required=""
        >
          <option
            value="Bungalow"
          >
            Bungalow
          </option>
          <option
            value="Cottage"
          >
            Cottage
          </option>
          <option
            value="End of Terrace"
          >
            End of Terrace
          </option>
          <option
            value="Flat"
          >
            Flat
          </option>
          <option
            value="Semi-detached"
          >
            Semi-detached
          </option>
          <option
            value="Terraced"
          >
            Terraced
          </option>
          <option
            disabled=""
            value=""
          >
             
          </option>
        </select>
      </label>
      <label
        for="bedrooms"
      >
        bedrooms 
        <input
          id="bedrooms"
          min="0"
          name="bedrooms"
          required=""
          step="1"
          type="number"
          value=""
        />
      </label>
      <label
        for="bathrooms"
      >
        bathrooms 
        <input
          id="bathrooms"
          min="0"
          name="bathrooms"
          required=""
          step="1"
          type="number"
          value=""
        />
      </label>
      <label
        for="price"
      >
        price Â£
        <input
          id="price"
          min="0"
          name="price"
          required=""
          step="1"
          type="number"
          value=""
        />
      </label>
      <label
        for="city"
      >
        city
        <select
          id="city"
          name="city"
          required=""
        >
          <option
            value="Leeds"
          >
            Leeds
          </option>
          <option
            value="Liverpool"
          >
            Liverpool
          </option>
          <option
            value="Manchester"
          >
            Manchester
          </option>
          <option
            value="Sheffield"
          >
            Sheffield
          </option>
          <option
            disabled=""
            value=""
          >
             
          </option>
        </select>
      </label>
      <div
        class="textbox"
      >
        <label
          class="textbox__label"
          for="email"
        >
          email
        </label>
        <input
          id="email"
          maxlength="64"
          name="email"
          required=""
          type="email"
          value=""
        />
      </div>
      <button
        type="submit"
      >
        Add
      </button>
    </form>
  </div>
</DocumentFragment>
`;
